### ISSUE TITLE
[unsafe-assignment-rule] handle multiple call signatures

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#128]
This applies to all `unsafe-assignment-rule` rules (mutable -> readonly, readonly -> mutable, optional property) but I'll frame this ticket around a mutable -> readonly example.

Consider this example, which is flagged by the rule that forbids mutable -> readonly assignment:

```ts
type MutableA = () => { a: string };
type ReadonlyA = () => { readonly a: string };

const ma: MutableA = () => { a: "" };
const ra: ReadonlyA = ma; // flagged as mutable -> readonly assignment due to the function return values
```

But what if those functions had multiple call signatures, like this?

```ts
        type MutableA = { a: string };
        type ReadonlyA = { readonly a: string };

        interface MutableFunc {
          (b: number): MutableA;
          (b: string): MutableA;
        }

        interface ReadonlyFunc {
          (b: number): ReadonlyA;
          (b: string): ReadonlyA;
        }
```

In that case the unsafe mutable -> readonly assignment is currently _not_ flagged by the rule.

```ts
        const mf: MutableFunc = (b: number | string): MutableA => {
          return { a: "" };
        };

        const rf: ReadonlyFunc = mf; // Not flagged; should be
```

The relevant TODO is this one, in unsafe-assignment-rule.ts:

```ts
        // eslint-disable-next-line functional/no-conditional-statement
        if (
          destinationCallSignatures.length > 1 ||
          sourceCallSignatures.length > 1
        ) {
          // TODO: work out how to handle multiple call signatures.
          // checker.getResolvedSignature() ?
          return false;
        }
```

And the test case that is considered valid but should be invalid is this one, in no-unsafe-mutable-readonly-assignment.test.ts:

```ts
    // mutable function return -> readonly function return (multiple call signatures).
    // TODO this should be invalid.
    {
      filename: "file.ts",
      code: `
        type MutableA = { a: string };
        type ReadonlyA = { readonly a: string };

        interface MutableFunc {
          (b: number): MutableA;
          (b: string): MutableA;
        }

        interface ReadonlyFunc {
          (b: number): ReadonlyA;
          (b: string): ReadonlyA;
        }

        const mf: MutableFunc = (b: number | string): MutableA => {
          return { a: "" };
        };

        const rf: ReadonlyFunc = mf;
      `,
    },
```
